lib/rucio/daemons/hermes/kafka/filters/default.py

```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+
+class default:
+    def __init__(self, logger, producer, topics):
+        self.logger = logger
+        self.producer = producer
+        self.topics = topics
+
+    def process(self, messages):
+        for topic in self.topics:
+            self._process(topic, messages)
```

Comment:
You aren't accumulating and returning the list of processed messages.  Doesn't that mean that they won't be deleted and will instead be re-sent?

Reply:
Yes, this is a bug in default, which I'll correct. This is kind of a catch-22.   If you blindly delete messages that were successfully sent, you run the risk of deleting messages on topics that failed to send.  As the code stands, it should delete messages only if they're sent on all topics, and send duplicates out if messages failed to be sent on any of the topics.  This isn't a long term solution.  Possibilities are:  restrict to one topic for Kafka in default, or store messages in the DB specific to each topic.  This isn't an issue for filtered_list.py, because messages are sent on specific topics (and not all of them), and if they aren't sent, they aren't deleted.

Reply:
The problem is that no messages are being deleted at all.  You're returning the list of `to_delete` from `_process()`, but then this code throws away that list rather than returning it as the result of `process()`.

I'm not exactly sure how this is interacting with the Kafka filter, but we are indeed getting hundreds of thousands of undeleted messages are piling up in the `messages` table.

Reply:
Yes, I understand the `to_delete` is an issue - what I was commenting on was that in addition to that problem, there's an additional problem of how to deal with messages sent to multiple topics, some of which can fail.  I'm going to create a new ticket to deal with that issue specifically.  I've added a temporary fix which addresses the `to_delete` list being thrown away.

The default filter doesn't interact with the Kafka filter in any way, and the hundreds of thousands of undeleted messages are either because HermesK wasn't running, or that additional services (like activemq) were configured, but not consumed properly by the base Rucio code.